[threaded pg] add assertion util to MultiThreadedTestCase

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90713
* #90712
* __->__ #90595

Differential Revision: [D42153662](https://our.internmc.facebook.com/intern/diff/D42153662)

XilunWu committed:
[threaded pg] have each thread use its own RNG

[ghstack-poisoned]

wanchaol commented on test/distributed/test_multi_threaded_pg.py:
```diff
@@ -127,6 +127,13 @@ def test_all_reduce(self):
+    def test_random_seed_consistency(self):
+        rng = torch.Generator()  # TODO: we need to hide this instantiation from test
+        self_tensor = torch.rand(3, 3, generator=rng)  # TODO: we need to hide generator=rng from test
+        rank_0_tensor = self_tensor.clone()
```
This is just to test whether rng works in multithreaded or not? That should already been tested by rng tests itself, we don't need to add a test like this.

XilunWu replied:
That's right. The goal I want to achieve is hiding line 131 and the explicit `generator=rng` in line 132 (otherwise we have to write this explicitly in every `rand()` call in tests). Currently this is the blocker. I'm digging a good solution. Any idea on it?

wanchaol replied:
Wouldn't just use `torch.set_rng_state` should work for this?

XilunWu replied:
If I understand correctly, `torch.set_rng.state` copies the passed state into `torch.default_generator` as its new internal state. However, this doesn't make each thread use its own RNG. The RNG they use is still the shared one. Right?

wanchaol reviewed (commented):
Can you add the actual functionality to MultithreadedTestCase directly and add tests to it instead?

XilunWu committed:
Update on "[threaded pg] have each thread use its own RNG"




[ghstack-poisoned]

XilunWu committed:
Update on "[threaded pg] have each thread use its own RNG"




[ghstack-poisoned]

XilunWu committed:
Update on "[threaded pg] have each thread use its own RNG"




[ghstack-poisoned]

XilunWu committed:
Update on "[threaded pg] have each thread use its own RNG"




[ghstack-poisoned]

XilunWu committed:
Update on "[threaded pg] have each thread use its own RNG"




[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42153662).

wanchaol commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -947,7 +947,7 @@ def threaded_run_test(self):
 
     def perThreadSetUp(self):
-        pass
+        super().setUp()  # TestCase.setUp() calls torch.manual_seed()
```
so is this PR more like syncing the initial seeds between threads? is it still necessary if we do `self.assertEqualRank0`?

XilunWu replied:
Yes this PR syncs initial seeds among CPU RNG and GPU RNG on each device. It's not necessary but may help prevent other failure in future.

XilunWu committed:
Update on "[threaded pg] have each thread use its own RNG"


Differential Revision: [D42153662](https://our.internmc.facebook.com/intern/diff/D42153662)

[ghstack-poisoned]

XilunWu committed:
Update on "[threaded pg] add assertion util to MultiThreadedTestCase"


Differential Revision: [D42153662](https://our.internmc.facebook.com/intern/diff/D42153662)

[ghstack-poisoned]

wanchaol commented on test/distributed/test_multi_threaded_pg.py:
```diff
@@ -4,15 +4,15 @@
 import torch
 import torch.distributed as dist
 from torch._C._distributed_c10d import ReduceOp
+from unittest import skip, SkipTest
```
nit: fix the lint.

wanchaol commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -956,6 +956,19 @@ def perThreadTearDown(self):
+    def rank(self) -> int:
+        return c10d.get_rank()
+
+    def assertEqualRank0(self, x, y):
```
can we make this like `def assertEqualOnRank(self, x, y, rank=0)`?

Maybe also add a `assertNotEqualOnRank` too?

wanchaol reviewed (approved):
lgtm, please fix the lint, I have some additional comments about testing. Please address them if you can.

XilunWu committed:
Update on "[threaded pg] add assertion util to MultiThreadedTestCase"


Differential Revision: [D42153662](https://our.internmc.facebook.com/intern/diff/D42153662)

[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42153662).

XilunWu commented:
@pytorchmergebot merge -g
